src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1698,18 +1698,150 @@ private static bool CheckNotNamespaceOrType(BoundExpression expr, DiagnosticBag
+        }
+
+        /// <summary>
+        /// Find the Deconstruct method for the expression on the right, that will fit the assignable bound expressions on the left.
+        /// Returns true if the Deconstruct method is found. If so, it outputs the method and all of its parameter types.
+        /// </summary>
+        private static bool FindDeconstruct(ImmutableArray<BoundExpression> checkedVariables, BoundExpression boundRHS, out MethodSymbol deconstructMethod, out ImmutableArray<TypeSymbol> parameterTypes, SyntaxNode node, DiagnosticBag diagnostics)
+        {
+            const string Deconstruct = "Deconstruct";
+
+            // find symbol for Deconstruct member
+            ImmutableArray<Symbol> candidates = boundRHS.Type.GetMembers(Deconstruct);
+            switch (candidates.Length)
+            {
+                case 0:
+                    Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
+                    goto error;
+                case 1:
+                    break;
+                default:
+                    Error(diagnostics, ErrorCode.ERR_AmbiguousDeconstruct, boundRHS.Syntax, boundRHS.Type);
+                    goto error;
+            }
+
+            Symbol deconstructMember = candidates[0];
+
+            // check that the deconstruct fits
+            if (deconstructMember.Kind != SymbolKind.Method)
+            {
+                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
+                goto error;
+            }
+
+            deconstructMethod = (MethodSymbol)deconstructMember;
+            if (deconstructMethod.MethodKind != MethodKind.Ordinary)
+            {
+                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
+                goto error;
+            }
+
+            if (deconstructMethod.ParameterCount != checkedVariables.Length)
+            {
+                Error(diagnostics, ErrorCode.ERR_DeconstructWrongParams, boundRHS.Syntax, boundRHS.Type);
+                goto error;
+            }
+
+            // get the parameter types
+            var parameters = deconstructMethod.Parameters;
+            var parameterTypesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(parameters.Length);
+            foreach (var parameter in parameters)
+            {
+                if (parameter.RefKind != RefKind.Out)
+                {
+                    Error(diagnostics, ErrorCode.ERR_DeconstructRequiresOutParams, boundRHS.Syntax, boundRHS.Type);
+                    goto error;
+                }
+
+                parameterTypesBuilder.Add(parameter.Type);
+            }
+
+            parameterTypes = parameterTypesBuilder.ToImmutableAndFree();
+
+            return true;
+
+            error:
+
+            deconstructMethod = new ErrorMethodSymbol(boundRHS.Type, ErrorTypeSymbol.UnknownResultType, Deconstruct);
+            return false;
```

Comment:
Consider setting `deconstructMethod` to null rather than `ErrorMethodSymbol` since the caller does not use the value in the error case. Then the `Deconstruct` `const` can be inlined in the one other use.